### DISCUSSION TITLE
Trying to create bundle on Ubuntu 2204 "Failed to load the AIR Runtime"

### DISCUSSION THREAD

**Discussion by rnmixon** [airsdk/Adobe-Runtime-Support#2419] [Q&A] [answered]
I'm using the latest apache Flex SDK and latest (5 .1.1.2) version of the Harman SDK (Flex version) and get this error. I've included at the end of this post a snippet from the run. The snippet  shows the command being issued and the error, as well as the toolkits I'm using. 

Background:

1. I expanded the apache-flex-sdk-4.16.1-bin folder from the download of the same name. Then expanded the AIRSDK_Flex_Linux_50.1.1.2.zip archive into the root of the apache-flex-sdk-4.16.1-bin folder.
2. This approach seems to work fine on both Windows and MacOS.

```
n2kops@dev2n2k:~/tools/air_projects/CountDownSimple$ ./build.sh CountDown
Environment variable AIR_HOME is set to: /home/n2kops/tools/apache-flex-sdk-4.16.1-bin
Flex/Air sdk at: "/home/n2kops/tools/apache-flex-sdk-4.16.1-bin"
Processing captive "CountDown"
AIR_HOME=/home/n2kops/tools/apache-flex-sdk-4.16.1-bin
+ /home/n2kops/tools/apache-flex-sdk-4.16.1-bin/bin/adt -package -keystore ../certs/sampleCert.pfx -storetype pkcs12 -storepass abc123def456 -tsa none -target bundle dist/linux CountDown-app.xml CountDown.swf
Failed to load the AIR Runtime
n2kops@dev2n2k:~/tools/air_projects/CountDownSimple$ ls -ld ~/tools/AIRSDK_Flex_Linux_50.1.1.2.zip ~/tools/apache-flex-sdk-4.16.1-bin
-rw-r--r--  1 n2kops n2kops 165402924 Jan  5 22:09 /home/n2kops/tools/AIRSDK_Flex_Linux_50.1.1.2.zip
drwxrwxr-x 16 n2kops n2kops      4096 Jan  5 23:21 /home/n2kops/tools/apache-flex-sdk-4.16.1-bin
n2kops@dev2n2k:~/tools/air_projects/CountDownSimple$
```

Any help or ideas are much appreciated.

Thank you - Richard

**Comment by pol2095** [comment]
You can install Flex SDK on Ubuntu using Apache Ant, see [here](https://forum.starling-framework.org/d/22177-adobe-flash-builder-4-7-new-air-and-flex-sdks/12)
replace "USERNAME" in "local.properties"
for the last Air SDK :
`ant -f installer.xml -Dflash.sdk.version=50.1 -Dair.sdk.version=50.1`

**Reply by rnmixon** [reply]
Sorry, I'm confused. How does installing AIR sdk 33.1 with ant fix my "Failed to load the AIR Runtime" error? I'm using Harman AIR 5.1.1.2.

Thank you for your help!

**Reply by pol2095** [reply]
For me, Flex SDK work using Apache Ant. The link is an example for 33.1 but it work with 50.1.1.2 too.
use `ant -f installer.xml -Dflash.sdk.version=50.1 -Dair.sdk.version=50.1`

![flex_linux](https://user-images.githubusercontent.com/6962232/212163428-e8f73b6e-26d4-4797-9f40-19f20dd7a572.png)

**Comment by ajwfrost** [comment, marked as answer]
For "Failed to load the AIR Runtime" - can you check that AIR has all the dependencies it needs? e.g. if you do:
```
ldd sdk_folder/runtimes/air/linux/Adobe\ AIR/Versions/1.0/libCore.so
```
it should highlight any dependent libraries that are "not found"..?

Can I also check you have a license file `adt.lic` in your `~/.airsdk/` folder?

thanks

**Comment by rnmixon** [comment]
Ahh! Thanks @ajwfrost  . The ldd command showed a lot of X dependencies missing - which makes sense, the build VM is headless (no X desktop installed).  There is also not adt.lic file (or even the ~/.airsdk folder) you referred to.

Some questions on how to get past this issue:

1. Can you recommend a minimal X desktop to install on Ubuntu 22.04? Trying to keep the build machine skinny.

2. Neither the Windows, MacOs or Linux environments have a "~/.airsdk" folder. The build works fine on Windows and MacOs. Is this specific to Linux? We have a Linux SDK (company name Maretron) - would that have included the adt.lic file?

3. Is it possible to build a bundle for all three environments (Windows, MacOs and Linux) from just the Windows SDK environment?

BTW, I'm not an AIR developer,  just a sysop putting together the production build process.

Thank you much - Richard

**Reply by ajwfrost** [reply]
Hi

We haven't really looked into this much - we've just been using normal Ubuntu desktop installations (or VMs). But given you're trying to set up a build machine .. the failure is happening when it's doing the "native product" conversion process which is itself an AIR application, but it's one that also runs without UI. So we may be able to do some changes so that these dependencies are switched to dynamic ones that are only needed if it's a graphical app....

The .airsdk folder is one where we put configuration files but also license files - and the use of AIR on Linux requires a license.. so if you'd purchased this via the website, we should have emailed you an adt.lic file that you can then put there manually.

We have been asked before about creating a bundle for other environments from each of them.. there are some OS-specific activities that happen though, particularly on Windows with some process editor stuff to change the "template" executable information. I think it's actually simpler on Linux and macOS as we have generic (text-config) ways of doing this, so it's something we had wondered about. But it's not been implemented yet..

Hope that helps!

thanks

**Reply by rnmixon** [reply]
Thank you for the detailed answers. It is clear what we need to do now.

My takeaway is that (for now) to build a captive bundle (only option on Linux) we need to run the build on the specific platform that the bundle will need execute on.

If a  decision is made to allow creating bundles for more than one platform, it is likely that either the Linus or MacOs environment would be addressed first - Windows support would be a bit more difficult.

Thank you much!

**Reply by rnmixon** [reply]
A quick followup in case it helps someone else with this issue.

After installing a minimal desktop, "sudo apt install ubuntu-desktop-minimal" all of the dependencies for the following command resolved (output at very end of this post):
- ldd /home/n2kops/tools/apache-flex-sdk-4.16.1-bin/runtimes/air/linux/Adobe\ AIR/Versions/1.0/libCore.so

Still, running my build from a non-X terminal resulted in an error:
```
n2kops@dev2n2k:~/tools/air_projects/CountDownSimple$ ./build.sh CountDown
Environment variable AIR_HOME is set to: /home/n2kops/tools/apache-flex-sdk-4.16.1-bin
Flex/Air sdk at: "/home/n2kops/tools/apache-flex-sdk-4.16.1-bin"
Processing captive "CountDown"
AIR_HOME=/home/n2kops/tools/apache-flex-sdk-4.16.1-bin
+ /home/n2kops/tools/apache-flex-sdk-4.16.1-bin/bin/adt -package -keystore ../certs/sampleCert.pfx -storetype pkcs12 -storepass abc123def456 -tsa none -target bundle dist/linux CountDown-app.xml CountDown.swf
unexpected failure: Platform conversion failed. Process exited with error 1. Output was "
(/home/n2kops/tools/apache-flex-sdk-4.16.1-bin/lib/nai/bin/naip:30153): Gtk-WARNING **: cannot open display:
"
...
```

From the VM console I ran "startx" and after basic setup was able to run the build successfully:
![SuccessfulLinuxBuild](https://user-images.githubusercontent.com/3441188/212549912-50fdb82a-f413-4b30-bc2a-a1ed36ede495.png)

Next to try a more complex AIR application.

Thanks again - Richard

Output from ldd command described above:
```
n2kops@dev2n2k:~/tools/air_projects/CountDownSimple$ ldd /home/n2kops/tools/apache-flex-sdk-4.16.1-bin/runtimes/air/linu
x/Adobe\ AIR/Versions/1.0/libCore.so
        linux-vdso.so.1 (0x00007ffcf954c000)
        libgtk-x11-2.0.so.0 => /lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0 (0x00007f55f0f89000)
        libgdk-x11-2.0.so.0 => /lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0 (0x00007f55f0ec8000)
        libgdk_pixbuf-2.0.so.0 => /lib/x86_64-linux-gnu/libgdk_pixbuf-2.0.so.0 (0x00007f55f2cc9000)
        libgobject-2.0.so.0 => /lib/x86_64-linux-gnu/libgobject-2.0.so.0 (0x00007f55f0e68000)
        libglib-2.0.so.0 => /lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007f55f0d2e000)
        libgthread-2.0.so.0 => /lib/x86_64-linux-gnu/libgthread-2.0.so.0 (0x00007f55f2cc2000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f55f2cbd000)
        libX11.so.6 => /lib/x86_64-linux-gnu/libX11.so.6 (0x00007f55f0bee000)
        libXcursor.so.1 => /lib/x86_64-linux-gnu/libXcursor.so.1 (0x00007f55f2cb1000)
        libfontconfig.so.1 => /lib/x86_64-linux-gnu/libfontconfig.so.1 (0x00007f55f0ba4000)
        libfreetype.so.6 => /lib/x86_64-linux-gnu/libfreetype.so.6 (0x00007f55f0adc000)
        libpango-1.0.so.0 => /lib/x86_64-linux-gnu/libpango-1.0.so.0 (0x00007f55f0a75000)
        libXrender.so.1 => /lib/x86_64-linux-gnu/libXrender.so.1 (0x00007f55f2ca2000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f55f0a59000)
        libxml2.so.2 => /lib/x86_64-linux-gnu/libxml2.so.2 (0x00007f55f0877000)
        libssl3.so => /lib/x86_64-linux-gnu/libssl3.so (0x00007f55f080f000)
        libnss3.so => /lib/x86_64-linux-gnu/libnss3.so (0x00007f55f06e2000)
        libnspr4.so => /lib/x86_64-linux-gnu/libnspr4.so (0x00007f55f06a2000)
        libsmime3.so => /lib/x86_64-linux-gnu/libsmime3.so (0x00007f55f0678000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f55f0673000)
        libGL.so.1 => /lib/x86_64-linux-gnu/libGL.so.1 (0x00007f55f05ec000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f55f03c2000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f55f02db000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f55f02bb000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f55f02b6000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f55f008e000)
        libgmodule-2.0.so.0 => /lib/x86_64-linux-gnu/libgmodule-2.0.so.0 (0x00007f55f0087000)
        libpangocairo-1.0.so.0 => /lib/x86_64-linux-gnu/libpangocairo-1.0.so.0 (0x00007f55f0075000)
        libXfixes.so.3 => /lib/x86_64-linux-gnu/libXfixes.so.3 (0x00007f55f006d000)
        libatk-1.0.so.0 => /lib/x86_64-linux-gnu/libatk-1.0.so.0 (0x00007f55f0043000)
        libcairo.so.2 => /lib/x86_64-linux-gnu/libcairo.so.2 (0x00007f55eff1b000)
        libgio-2.0.so.0 => /lib/x86_64-linux-gnu/libgio-2.0.so.0 (0x00007f55efd43000)
        libpangoft2-1.0.so.0 => /lib/x86_64-linux-gnu/libpangoft2-1.0.so.0 (0x00007f55efd28000)
        libXinerama.so.1 => /lib/x86_64-linux-gnu/libXinerama.so.1 (0x00007f55efd23000)
        libXi.so.6 => /lib/x86_64-linux-gnu/libXi.so.6 (0x00007f55efd0d000)
        libXrandr.so.2 => /lib/x86_64-linux-gnu/libXrandr.so.2 (0x00007f55efd00000)
        libXcomposite.so.1 => /lib/x86_64-linux-gnu/libXcomposite.so.1 (0x00007f55efcfb000)
        libXdamage.so.1 => /lib/x86_64-linux-gnu/libXdamage.so.1 (0x00007f55efcf6000)
        libXext.so.6 => /lib/x86_64-linux-gnu/libXext.so.6 (0x00007f55efce1000)
        libpng16.so.16 => /lib/x86_64-linux-gnu/libpng16.so.16 (0x00007f55efca6000)
        libjpeg.so.8 => /lib/x86_64-linux-gnu/libjpeg.so.8 (0x00007f55efc23000)
        libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x00007f55efc16000)
        libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f55efba0000)
        libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f55efb76000)
        libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00007f55efb45000)
        libuuid.so.1 => /lib/x86_64-linux-gnu/libuuid.so.1 (0x00007f55efb3a000)
        libbrotlidec.so.1 => /lib/x86_64-linux-gnu/libbrotlidec.so.1 (0x00007f55efb2c000)
        libfribidi.so.0 => /lib/x86_64-linux-gnu/libfribidi.so.0 (0x00007f55efb10000)
        libthai.so.0 => /lib/x86_64-linux-gnu/libthai.so.0 (0x00007f55efb05000)
        libharfbuzz.so.0 => /lib/x86_64-linux-gnu/libharfbuzz.so.0 (0x00007f55efa36000)
        libicuuc.so.70 => /lib/x86_64-linux-gnu/libicuuc.so.70 (0x00007f55ef839000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f55ef80e000)
        libnssutil3.so => /lib/x86_64-linux-gnu/libnssutil3.so (0x00007f55ef7dc000)
        libplc4.so => /lib/x86_64-linux-gnu/libplc4.so (0x00007f55ef7d5000)
        libplds4.so => /lib/x86_64-linux-gnu/libplds4.so (0x00007f55ef7d0000)
        libGLdispatch.so.0 => /lib/x86_64-linux-gnu/libGLdispatch.so.0 (0x00007f55ef716000)
        libGLX.so.0 => /lib/x86_64-linux-gnu/libGLX.so.0 (0x00007f55ef6e2000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f55f2d09000)
        libpixman-1.so.0 => /lib/x86_64-linux-gnu/libpixman-1.so.0 (0x00007f55ef637000)
        libxcb-shm.so.0 => /lib/x86_64-linux-gnu/libxcb-shm.so.0 (0x00007f55ef632000)
        libxcb-render.so.0 => /lib/x86_64-linux-gnu/libxcb-render.so.0 (0x00007f55ef623000)
        libmount.so.1 => /lib/x86_64-linux-gnu/libmount.so.1 (0x00007f55ef5dd000)
        libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007f55ef5b1000)
        libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x00007f55ef5ab000)
        libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f55ef5a3000)
        libbrotlicommon.so.1 => /lib/x86_64-linux-gnu/libbrotlicommon.so.1 (0x00007f55ef580000)
        libdatrie.so.1 => /lib/x86_64-linux-gnu/libdatrie.so.1 (0x00007f55ef575000)
        libgraphite2.so.3 => /lib/x86_64-linux-gnu/libgraphite2.so.3 (0x00007f55ef54e000)
        libicudata.so.70 => /lib/x86_64-linux-gnu/libicudata.so.70 (0x00007f55ed930000)
        libblkid.so.1 => /lib/x86_64-linux-gnu/libblkid.so.1 (0x00007f55ed8f9000)
        libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007f55ed862000)
        libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007f55ed848000)
        libmd.so.0 => /lib/x86_64-linux-gnu/libmd.so.0 (0x00007f55ed83b000)
n2kops@dev2n2k:~/tools/air_projects/CountDownSimple$
```
